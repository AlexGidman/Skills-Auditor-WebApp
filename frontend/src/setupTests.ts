// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
