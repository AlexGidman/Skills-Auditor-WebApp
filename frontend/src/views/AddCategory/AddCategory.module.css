.content {
    margin-top: 1rem;
    align-items: center;
}

.textField {
    min-width: 10rem;
    margin: 0 3%;
    flex: 60%;
}

.button {
    margin: 3%;
    flex: 44%;
    min-width: 14rem;
}

.form {
    display: flex;
    flex-wrap: wrap;
    margin: 0 10%;
}

.item {
    min-width: 10rem;
    margin: 0 3%;
    flex: 44%;
}

@media (max-width: 900px) {
    .form {
        margin: 0;
    }
}
