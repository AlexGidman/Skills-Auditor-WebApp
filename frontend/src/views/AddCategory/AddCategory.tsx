import React, { useState, useEffect } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import { useImmer } from "use-immer";

import { Button } from "../../components/Button/Button";
import { TextInput } from "../../components/TextInput/TextInput";
import { useAPI } from "../../utility/helper";

import styles from "./AddCategory.module.css";
import { addCategory } from "../../utility/apiRequests";
import cx from "classnames";

export const AddCategory = () => {
    return (
        <>
            <header className={styles.content}>
                <h1>Add Category</h1>
            </header>
            <section>
                <Form />
            </section>
        </>
    );
};

const Form = () => {
    const [inputValues, setInputValues] = useImmer({
        name: "",
    });

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const { name, value } = e.target;
        setInputValues((prev) => ({
            ...prev,
            [name]: value,
        }));
    };

    const navigate = useNavigate();
    // @ts-ignore TODO fix type for AppOutletContext here
    const [, setShowToast] = useOutletContext();

    const [formChanged, setFormChanged] = useState(false);
    const [navigateBack, setNavigateBack] = useState(false);

    const {
        data: addData,
        error,
        loading,
        callback: executeAddCategory,
    } = useAPI(addCategory, Object.values(inputValues), true);

    useEffect(() => {
        setFormChanged(inputValues.name !== "" && inputValues.name !== undefined);
    }, [inputValues]);

    useEffect(() => {
        if (navigateBack && !loading) {
            addData && setShowToast({ success: "Successfully added category" });
            error && setShowToast({ error: "Failed to add category" });
            navigate(-1);
        }
    }, [error, loading, navigate, navigateBack, setShowToast, addData]);

    return (
        <form className={styles.form}>
            <TextInput
                labelText="Enter a category name"
                className={cx(styles.textField, styles.content)}
                value={inputValues.name}
                name="name"
                onChange={handleChange}
                required
            />

            <Button
                className={styles.button}
                loading={loading}
                disabled={!formChanged}
                onClick={(e) => {
                    e.preventDefault();
                    executeAddCategory();
                    setNavigateBack(true);
                }}
            >
                Add Category
            </Button>

            <Button
                className={styles.button}
                onClick={(e) => {
                    e.preventDefault();
                    navigate(-1);
                }}
            >
                Cancel
            </Button>
        </form>
    );
};
