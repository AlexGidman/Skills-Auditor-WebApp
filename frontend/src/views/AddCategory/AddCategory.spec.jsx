import { render, screen } from "@testing-library/react";
import { mockAdminUser, mockError } from "../../setupTests";
import { AddCategory } from "./AddCategory";
import * as ReactDom from "react-router-dom";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
    };
});

describe("AddCategory", () => {
    it("renders correctly when data successful", async () => {
        ReactDom.useOutletContext.mockReturnValue([mockAdminUser, jest.fn()]);
        const { container } = render(
            <ReactDom.MemoryRouter>
                <AddCategory />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "Add Category" })).toBeInTheDocument();
        expect(screen.getByRole("textbox", { name: "Enter a category name:" })).toBeInTheDocument();
        expect(screen.getByRole("textbox", { name: "Enter a category name:" })).toHaveValue("");

        expect(screen.getByRole("button", { name: "Add Category" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Cancel" })).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
