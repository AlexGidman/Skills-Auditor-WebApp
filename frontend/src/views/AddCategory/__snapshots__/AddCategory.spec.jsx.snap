// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory renders correctly when data successful 1`] = `
<div>
  <header
    class="content"
  >
    <h1>
      Add Category
    </h1>
  </header>
  <section>
    <form
      class="form"
    >
      <div
        class="component textField content"
      >
        <label
          class="label"
          for="Enter-a category name-undefined"
        >
          Enter a category name
          :
        </label>
        <input
          class="input"
          id="Enter-a category name-undefined"
          maxlength="100"
          minlength="0"
          name="name"
          pattern="[a-zA-ZÀ-ȕ-]+$"
          required=""
          value=""
        />
      </div>
      <button
        class="component button"
        disabled=""
      >
        Add Category
      </button>
      <button
        class="component button"
      >
        Cancel
      </button>
    </form>
  </section>
</div>
`;
